### PR TITLE
User refactoring

### PR DESCRIPTION
**Pull request by rufus1303**
Implementing changes and adding tests for User handling

### CODE REVIEW

File: src/main/java/spring/workshop/expenses/rest/UserController.java [hunk 1]
```diff
@@ -24,14 +27,23 @@ public class MainController {
...
+    // @RequestParam means it is a parameter from the DELETE request
+
+    User entitytodelete = userRepository.findByName(name);
+    userRepository.delete(entitytodelete);
+    return "Deleted";
+  }
+
   @GetMapping(path="/all")
-  public @ResponseBody Iterable<User> getAllUsers() {
+  public @ResponseBody List<User> getAllUsers() {
```
Comment (victoria-marano) [inline src/main/java/spring/workshop/expenses/rest/UserController.java:46]: Can you also please add an endpoint that returns one user once by id and once by name so we can utilize it later in other features?

File: src/main/java/spring/workshop/expenses/entities/User.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
 @Table(name="user_tab")
 public class User {
 
-  @Id
+@Id
```
Comment (victoria-marano) [inline src/main/java/spring/workshop/expenses/entities/User.java:15]: this looks a bit off, did you activate the format on save in VSCode?

File: src/main/java/spring/workshop/expenses/entities/User.java [hunk 3]
```diff
@@ -21,6 +21,13 @@ public class User {
   @NonNull
   private String name;
 
+    public User() {
+    }
+
+    public User(String name) {
```
Comment (victoria-marano) [inline src/main/java/spring/workshop/expenses/entities/User.java:27]:
```suggestion
    public User(@NonNull String name) {
```

File: src/main/java/spring/workshop/expenses/rest/UserController.java [hunk 4]
```diff
@@ -24,14 +27,23 @@ public class MainController {
...
     userRepository.save(n);
     return "Saved";
   }
 
+  @DeleteMapping(path="/delete") // Map ONLY DELETE Requests
+  public @ResponseBody String deleteUser (@RequestParam String name) {
+    // @ResponseBody means the returned String is the response, not a view name
+    // @RequestParam means it is a parameter from the DELETE request
+
+    User entitytodelete = userRepository.findByName(name);
```
Comment (victoria-marano) [inline src/main/java/spring/workshop/expenses/rest/UserController.java:40]: what happens when user is not found?

File: src/test/java/spring/workshop/expenses/integration/UserControllerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        
+        // Send a GET request to retrieve all users
+        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
+        
+        // Assert HTTP status code is OK
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+        
+        // Convert response body to a JSONArray and validate its length
+        // Assuming the response body is in JSON format representing an array of users
+        assertEquals(1, new JSONArray(response.getBody()).length());
```
Comment (victoria-marano) [inline src/test/java/spring/workshop/expenses/integration/UserControllerIntegrationTests.java:105]: Does this pass? in the test env there are 5 users already in the db, I wonder how this works!

File: src/test/java/spring/workshop/expenses/unit/UserControllerTests.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+@SpringBootTest
+@Transactional
+@Rollback
+class UserControllerTests {
+
+    @Autowired
+    private UserController userController;
+
+    @Autowired
+    private UserRepository userRepository;
```
Comment (victoria-marano) [inline src/test/java/spring/workshop/expenses/unit/UserControllerTests.java:28]: This stops being a unit test when you include two real components (UserController  and UserRepository) , you only test one component at a time in unit tests and you mock the rest.

File: src/main/java/spring/workshop/expenses/entities/User.java (hunk 2 above)
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/entities/User.java:15]: Now I did. Changed the formatting

File: src/main/java/spring/workshop/expenses/rest/UserController.java (hunk 4 above)
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/rest/UserController.java:40]: Added an error handling for this case. Also added an error handling for /add method.

File: src/test/java/spring/workshop/expenses/unit/UserControllerTests.java (hunk 6 above)
Comment (rufus1303) [reply src/test/java/spring/workshop/expenses/unit/UserControllerTests.java:28]: I still have to work on that...

File: src/test/java/spring/workshop/expenses/integration/UserControllerIntegrationTests.java (hunk 5 above)
Comment (rufus1303) [reply src/test/java/spring/workshop/expenses/integration/UserControllerIntegrationTests.java:105]: This comment belongs to testGetAllUsers(). I changed it but the check in the end is a little bit simple.

Comment (rufus1303) [reply src/test/java/spring/workshop/expenses/integration/UserControllerIntegrationTests.java:105]: Added tests testGetUserById() and testGetUserByName(). The do not work yet.

File: src/main/java/spring/workshop/expenses/rest/UserController.java (hunk 1 above)
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/rest/UserController.java:46]: Endpoints added.

File: src/test/java/spring/workshop/expenses/unit/UserControllerTests.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.test.annotation.Rollback;
+
+import jakarta.transaction.Transactional;
+import spring.workshop.expenses.entities.User;
+import spring.workshop.expenses.repos.UserRepository;
+import spring.workshop.expenses.rest.UserController;
+
+@SpringBootTest
+@Transactional
+@Rollback
```
Comment (victoria-marano) [inline src/test/java/spring/workshop/expenses/unit/UserControllerTests.java:22]: @ActiveProfiles("test") missing here

File: src/main/java/spring/workshop/expenses/repos/UserRepository.java [hunk 8]
```diff
@@ -1,12 +1,22 @@
...
 
-public interface UserRepository  extends CrudRepository<User, Integer> {
-    
+public interface UserRepository extends JpaRepository<User, Integer> {
+
+    Optional<User> findByName(String name);
+
+    Optional<User> findById(Integer id);
+
+    @Transactional
```
[the remark this answers is not in the document]
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/repos/UserRepository.java:19]: "even though all repo methods are transactional, newly declared methods are not by default. deleteById() is inhereted from CrudRepository and is therefore transactional
annotation can stay in the interface"

File: src/main/java/spring/workshop/expenses/rest/UserController.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import spring.workshop.expenses.entities.User;
+import spring.workshop.expenses.services.UserService;
+
+@RestController
+@RequestMapping(path = "/user")
```
[the remark this answers is not in the document]
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/rest/UserController.java:24]: ok

File: src/main/java/spring/workshop/expenses/rest/UserController.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+public class UserController {
+
+  private UserService userService;
```
[the remark this answers is not in the document]
Comment (rufus1303) [reply src/main/java/spring/workshop/expenses/rest/UserController.java:27]: Added Autowired annotation for the field injection below

File: src/test/java/spring/workshop/expenses/integration/User/UserControllerIntegrationTests.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    @Test
+    public void testUpdateUser() throws Exception {
+
+        // Setting up request header and body for the PUT request
+        // Constructing the request body with the user to be updated
+        HttpHeaders requestHeader = new HttpHeaders();
+        requestHeader.setContentType(MediaType.APPLICATION_JSON);
+        String requestBody = "{\"id\": 300, \"name\": \"Test2\"}";
```
[the remark this answers is not in the document]
Comment (rufus1303) [reply src/test/java/spring/workshop/expenses/integration/User/UserControllerIntegrationTests.java:87]: ok

File: src/main/java/spring/workshop/expenses/rest/UserController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (victoria-marano) [reply src/main/java/spring/workshop/expenses/rest/UserController.java:28]: You mean constructor injection :)

**Review by victoria-marano — APPROVED** [review]

**Pull request merged**